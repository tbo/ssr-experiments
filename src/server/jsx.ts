import { Readable, PassThrough } from 'stream';

(global as any).createElement = (
  type: any,
  props: Record<string, any> | undefined,
  ...children: any[]
): Promise<Element> | Element => ({
  type,
  props: props || {},
  children,
});

interface Element<P = Record<string, any> | undefined, C = any> {
  type: ((props: P, context?: C) => Promise<Element<P, C> | string> | Element<P, C> | string) | string;
  props: P;
  children: (Element<P, C> | Promise<Element<P, C>>)[];
}
const CAMEL_CASE_PATTERN = new RegExp('[A-Z]*[a-z]+', 'g');

const toKebapCase = (value: string) => value.match(CAMEL_CASE_PATTERN)!.join('-');

const toString = (attribute: [string, any]) => {
  let [key, value] = attribute;
  if (value == null) {
    return '';
  } else if (key === 'className') {
    key = 'class';
  } else if (key === 'style') {
    value = Object.entries(value)
      .map((entry) => [toKebapCase(entry[0]), entry[1]].join(':'))
      .join(';')
      .toLowerCase();
  }
  return ` ${key.toLowerCase()}="${value}"`;
};

type Node = JSX.Element | Promise<Element | string> | Element | string;

type OutputEntry = string | Promise<Node>;

export const render = (RootComponent: (context: any) => Node, context?: any): Readable => {
  const outputQueue: OutputEntry[] = ['<!DOCTYPE html>'];
  const parseNode = (node: Node): OutputEntry[] => {
    if (typeof node === 'string') {
      return [node];
    } else if (typeof node === 'number') {
      return [String(node)];
    } else if (node instanceof Promise) {
      (node as Promise<Node>).then(parseNode).then((elements) => spliceNode(elements, outputQueue.indexOf(node)));
      return [node];
    } else if (Array.isArray(node)) {
      return node.flatMap(parseNode);
    }
    const { type, props, children } = node;
    if (typeof node.type === 'function') {
      if (!children) {
        return parseNode(node.type(node.props, context));
      }
      return parseNode(node.type({ ...node.props, children: (children as any).flatMap(parseNode) }, context));
    }
    const propString = props ? Object.entries(props).map(toString).join('') : '';
    if (!children) {
      return [`<${type}${propString}/>`];
    }
    return [`<${type}${propString}>`, ...(children as any).flatMap(parseNode), `</${type}>`];
  };
  const processQueue = () => {
    while (typeof outputQueue[0] === 'string') {
      sink.write(outputQueue[0]);
      outputQueue.shift();
    }
    if (outputQueue[0] instanceof Promise) {
      outputQueue[0].then(processQueue);
    } else {
      sink.end();
    }
  };
  const spliceNode = (elements: OutputEntry[], position: number) => {
    if (position !== -1) {
      outputQueue.splice(position, 1, ...elements);
    }
  };
  const sink = new PassThrough({});

  outputQueue.push(...parseNode(RootComponent(context)));
  processQueue();
  return sink;
};
