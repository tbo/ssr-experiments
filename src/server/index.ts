import fastify from 'fastify';
import pointOfView from 'point-of-view';
import handlebars from 'handlebars';
import fastifyStatic from 'fastify-static';
import path from 'path';
import { Client } from 'undici';
import { Stream } from 'stream';
import replaceStream from 'replacestream';
import compress from 'fastify-compress';
import multipart from 'fastify-multipart';
import StreamSSE from 'ssestream2';
import { readFileSync } from 'fs';
import Example from './components/example';
import { render } from './jsx';
import Category from './pages/category';

const timeTemplate = handlebars.compile(readFileSync('./src/templates/time.hbs', 'utf-8'));
const algolia = new Client('https://latency-dsn.algolia.net');

const getRange = (to: number) => [...Array(to).keys()];

const search = (query: string, page = 0): Promise<string | undefined> =>
  new Promise((resolve, reject) =>
    algolia.request(
      {
        path:
          '/1/indexes/*/queries?x-algolia-api-key=6be0576ff61c053d5f9a3225e2a90f76&x-algolia-application-id=latency',
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          requests: [{ indexName: 'ikea', params: `query=${query}&hitsPerPage=150&page=${page}` }],
        }),
      },
      (error: Error | undefined, { statusCode, body }: any) => {
        if (error || statusCode >= 400) {
          return reject(error);
        }
        body.setEncoding('utf8');
        let response = '';

        body.on('data', (chunk: string) => (response += chunk));
        body.on('end', () => resolve(response));
      },
    ),
  );

const app = fastify({ logger: false });

app
  .addHook('preHandler', function (request, reply, next) {
    if (request.headers.accept?.startsWith('text/event-stream') && request.raw.url !== '/time') {
      reply.header('content-type', 'text/event-stream').code(204).send('');
      return;
    }
    next();
  })
  .register(pointOfView, {
    engine: {
      handlebars,
    },
    options: {
      partials: {
        header: '/src/templates/header.hbs',
        footer: '/src/templates/footer.hbs',
      },
    },
  })
  .register(compress)
  .register(multipart, { addToBody: true })
  .register(fastifyStatic, { root: path.join(__dirname, '../assets'), prefix: '/assets' })
  .get('/', async (request, reply) => {
    const { query } = request.query;
    const params: Record<string, any> = { query };
    if (query) {
      const activePage = Number(request.query.page) || 0;
      const response = await search(request.query.query, activePage);
      if (response) {
        params.searchResult = JSON.parse(response).results[0];
        const pages = params.searchResult.nbPages;
        if (pages > 0) {
          params.pages = getRange(pages).map((page) => ({
            active: activePage === page,
            label: page + 1,
            url: `/?query=${query}&page=${page}`,
          }));
        }
      }
    }
    reply.view('/src/templates/search.hbs', params);
  })
  .get('/test', async (request, reply) => {
    return render(Example(), { request, reply });
  })
  .get('/products/:category', async (request, reply) => {
    return render(Category(), { request, reply });
  })
  .get('/post-search', (_request, reply) => {
    reply.view('/src/templates/post-search.hbs');
  })
  .post('/post-search', async (request, reply) => {
    const { query } = request.body;
    const params: Record<string, any> = { query };
    if (query) {
      const response = await search(query);
      if (response) {
        params.searchResult = JSON.parse(response).results[0];
      }
    }
    reply.view('/src/templates/post-search.hbs', params);
  })
  .get('/time', (request, reply) => {
    const getParams = () => ({ time: new Date() });
    if (request.headers.accept.startsWith('text/event-stream')) {
      reply.header('content-type', 'text/event-stream');
      const sse = new StreamSSE(request.raw, {
        'Access-Control-Allow-Origin': '*',
      });
      setInterval(
        () =>
          sse.write({
            data: timeTemplate(getParams()),
          }),
        1000,
      );
      sse.pipe(reply.res);
    } else {
      reply.view('/src/templates/time.hbs', getParams());
    }
  })
  .addHook('onSend', async (_request, reply, payload) => {
    if (reply.getHeader('Content-type')?.startsWith('text/html')) {
      const inject = '<script src="/assets/hybrid.js"></script>';
      reply.removeHeader('content-length');
      if (typeof payload === 'string') {
        return payload.replace('</body>', inject + '</body>');
      } else if (payload instanceof Stream) {
        return payload.pipe(replaceStream('</body>', inject + '</body>'));
      }
    }
    return payload;
  })
  .listen(3000)
  .catch((error) => {
    app.log.error(error);
    process.exit(1);
  });
