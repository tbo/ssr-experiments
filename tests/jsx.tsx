import { render } from '../src/server/jsx';
import { Readable } from 'stream';

const toString = async (stream: Readable) => {
  let buffer = '';
  stream.on('data', (data: string) => (buffer += data));
  return new Promise((resolve) => stream.on('end', () => resolve(buffer)));
};

const expectJSXtoMatchSnapshot = async (getComponent: () => JSX.Element) =>
  expect(await toString(render(getComponent, {}))).toMatchSnapshot();

describe('JSX middleware', () => {
  test('Render simple jsx tag', () => expectJSXtoMatchSnapshot(() => <div>simple component</div>));

  test('Render attributes', () =>
    expectJSXtoMatchSnapshot(() => (
      <div id="test" style={{ padding: 3, textAlign: 'right' }}>
        simple component
      </div>
    )));

  test('Do not render empty attribute values', () =>
    expectJSXtoMatchSnapshot(() => <div id={undefined}>simple component</div>));
});
