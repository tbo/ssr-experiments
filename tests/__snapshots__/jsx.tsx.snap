// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX middleware Do not render empty attribute values 1`] = `"<!DOCTYPE html><div>simple component</div>"`;

exports[`JSX middleware Handle 0 as valid child 1`] = `"<!DOCTYPE html><div>0</div>"`;

exports[`JSX middleware Handle 0 as valid child 2`] = `"<!DOCTYPE html><div>0</div>"`;

exports[`JSX middleware Handle atomic values 1`] = `"<!DOCTYPE html>only text"`;

exports[`JSX middleware Handle atomic values 2`] = `"<!DOCTYPE html>42"`;

exports[`JSX middleware Handle atomic values 3`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle atomic values 4`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle atomic values 5`] = `"<!DOCTYPE html>only text"`;

exports[`JSX middleware Handle atomic values 6`] = `"<!DOCTYPE html>42"`;

exports[`JSX middleware Handle atomic values 7`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle atomic values 8`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle empty children 1`] = `"<!DOCTYPE html><div/>"`;

exports[`JSX middleware Handle empty children 2`] = `"<!DOCTYPE html><div/>"`;

exports[`JSX middleware Handle empty children 3`] = `"<!DOCTYPE html><div/>"`;

exports[`JSX middleware Handle empty children 4`] = `"<!DOCTYPE html><div></div>"`;

exports[`JSX middleware Handle empty children 5`] = `"<!DOCTYPE html><div></div>"`;

exports[`JSX middleware Handle lists 1`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle lists 2`] = `"<!DOCTYPE html>42text<p key=\\"jsx\\">jsx</p>"`;

exports[`JSX middleware Handle lists 3`] = `"<!DOCTYPE html>"`;

exports[`JSX middleware Handle lists 4`] = `"<!DOCTYPE html>42text<p key=\\"jsx\\">jsx</p>"`;

exports[`JSX middleware Parallelize execution order 1`] = `"<!DOCTYPE html><div id=\\"abc\\" style=\\"color:red;padding:3\\"><h1>JSX Example</h1><p>some text</p>"`;

exports[`JSX middleware Parallelize execution order 2`] = `"<!DOCTYPE html><div id=\\"abc\\" style=\\"color:red;padding:3\\"><h1>JSX Example</h1><p>some text</p><p class=\\"SomeComponent\\"><span>123</span><div><b>A bold message</b><span>Async string</span></div></p></div>"`;

exports[`JSX middleware Render attributes 1`] = `"<!DOCTYPE html><div id=\\"test\\" style=\\"padding:3;text-align:right\\" data-testid=\\"testid\\">simple component</div>"`;

exports[`JSX middleware Render custom components with children 1`] = `"<!DOCTYPE html><div><span>Children</span></div>"`;

exports[`JSX middleware Render custom components with children 2`] = `"<!DOCTYPE html><span><b>A bold message</b></span>"`;

exports[`JSX middleware Render custom components with children 3`] = `"<!DOCTYPE html><span/>"`;

exports[`JSX middleware Render custom components with children 4`] = `"<!DOCTYPE html><span>42</span>"`;

exports[`JSX middleware Render custom components with properties 1`] = `"<!DOCTYPE html><div><span title=\\"A nice title\\">42</span></div>"`;

exports[`JSX middleware Render custom components with properties 2`] = `"<!DOCTYPE html><span title=\\"Another title\\">0</span>"`;

exports[`JSX middleware Render custom components with properties 3`] = `"<!DOCTYPE html><div><span title=\\"A nice title\\">42</span></div>"`;

exports[`JSX middleware Render custom components with properties 4`] = `"<!DOCTYPE html><span title=\\"Another title\\">0</span>"`;

exports[`JSX middleware Render custom components without properties 1`] = `"<!DOCTYPE html><div><span>custom component</span></div>"`;

exports[`JSX middleware Render custom components without properties 2`] = `"<!DOCTYPE html><span>custom component</span>"`;

exports[`JSX middleware Render custom components without properties 3`] = `"<!DOCTYPE html><div><span>async custom component</span></div>"`;

exports[`JSX middleware Render custom components without properties 4`] = `"<!DOCTYPE html><span>async custom component</span>"`;

exports[`JSX middleware Render inline JavaScript 1`] = `"<!DOCTYPE html><ul><li key=\\"a\\">a</li><li key=\\"b\\">b</li><li key=\\"c\\">c</li></ul>"`;

exports[`JSX middleware Render inline JavaScript 2`] = `"<!DOCTYPE html><ul><li key=\\"a\\">a</li><li key=\\"b\\">b</li><li key=\\"c\\">c</li></ul>"`;

exports[`JSX middleware Render multiple nested elements 1`] = `"<!DOCTYPE html><div><span>first</span>second<span><p>third</p></span>fourth</div>"`;

exports[`JSX middleware Render nested elements 1`] = `"<!DOCTYPE html><div><span>nested component</span></div>"`;

exports[`JSX middleware Render simple jsx tag 1`] = `"<!DOCTYPE html><div>simple component</div>"`;

exports[`JSX middleware Trigger and stream components eagerly 1`] = `"<!DOCTYPE html><div><div>Leaf A</div><div>"`;

exports[`JSX middleware Trigger and stream components eagerly 2`] = `"<!DOCTYPE html><div><div>Leaf A</div><div><div>Leaf B</div>"`;

exports[`JSX middleware Trigger and stream components eagerly 3`] = `"<!DOCTYPE html><div><div>Leaf A</div><div><div>Leaf B</div><div>"`;

exports[`JSX middleware Trigger and stream components eagerly 4`] = `"<!DOCTYPE html><div><div>Leaf A</div><div><div>Leaf B</div><div><div>Leaf D</div></div>"`;

exports[`JSX middleware Trigger and stream components eagerly 5`] = `"<!DOCTYPE html><div><div>Leaf A</div><div><div>Leaf B</div><div><div>Leaf D</div></div><div>Leaf E</div></div>"`;

exports[`JSX middleware Trigger and stream components eagerly 6`] = `"<!DOCTYPE html><div><div>Leaf A</div><div><div>Leaf B</div><div><div>Leaf D</div></div><div>Leaf E</div></div><div>Leaf F</div></div>"`;
